File: app/src/main/java/com/example/uos_lms/core/domain/model/ExamSchedule.java
package com.example.uos_lms.core.domain.model;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

@Getter
@Builder(toBuilder = true)
@EqualsAndHashCode
@ToString
public class ExamSchedule {
    @Builder.Default
    private final String id = "";
    @Builder.Default
    private final ExamType examType = ExamType.MID_TERM;
    @Builder.Default
    private final String subjectId = "";
    @Builder.Default
    private final String subjectCode = "";
    @Builder.Default
    private final String subjectTitle = "";
    @Builder.Default
    private final String departmentId = "";
    @Builder.Default
    private final String semesterId = "";
    @Builder.Default
    private final String teacherUid = "";
    @Builder.Default
    private final String teacherName = "";
    @Builder.Default
    private final String invigilatorUid = "";
    @Builder.Default
    private final String invigilatorName = "";
    @Builder.Default
    private final String room = "";
    private final long examDateMillis;
    @Builder.Default
    private final String examDateKey = "";
    private final int startTimeMinutes;
    private final int endTimeMinutes;
    @Builder.Default
    private final ExamScheduleStatus status = ExamScheduleStatus.DRAFT;
    @Builder.Default
    private final String createdBy = "";
    @Builder.Default
    private final long createdAt = 0L;
    private final Long publishedAt;
    private final Long lockedAt;
}

File: backend/src/models/ExamSchedule.js
const mongoose = require('mongoose');
const { Schema } = mongoose;

const examScheduleSchema = new Schema(
  {
    examType: { type: String, enum: ['MID_TERM', 'FINAL_TERM'], default: 'MID_TERM' },
    subjectId: { type: Schema.Types.ObjectId, ref: 'Subject', required: true },
    departmentId: { type: Schema.Types.ObjectId, ref: 'Department', required: true },
    semesterId: { type: Schema.Types.ObjectId, ref: 'Semester', required: true },
    teacherId: { type: String, ref: 'User', default: null },
    invigilatorId: { type: String, ref: 'User', default: null },
    room: { type: String, required: true, trim: true },
    roomNormalized: { type: String, required: true },
    examDate: { type: Date, required: true },
    examDateKey: { type: String, required: true },
    startTimeMinutes: { type: Number, required: true },
    endTimeMinutes: { type: Number, required: true },
    status: { type: String, enum: ['DRAFT', 'PUBLISHED', 'LOCKED'], default: 'DRAFT' },
    createdBy: { type: String, ref: 'User', required: true },
    publishedAt: { type: Date, default: null },
    lockedAt: { type: Date, default: null },
  },
  { timestamps: true }
);

// No cohort-uniqueness constraint - a department legitimately has many exams on the same
// day for different subjects (matches the pre-migration app). Only invigilator/room clashes
// are guarded.
examScheduleSchema.index(
  { invigilatorId: 1, examDateKey: 1, startTimeMinutes: 1 },
  { unique: true, partialFilterExpression: { invigilatorId: { $type: 'string' } }, name: 'invigilator_slot_unique' }
);
examScheduleSchema.index(
  { roomNormalized: 1, examDateKey: 1, startTimeMinutes: 1 },
  { unique: true, name: 'room_slot_unique' }
);

module.exports = mongoose.model('ExamSchedule', examScheduleSchema);
